pkg/relayer/session/session.go

```diff
@@ -191,27 +200,26 @@ func (rs *relayerSessionsManager) mapBlockToSessionsToClaim(
 						continue
 					}
 
-					sessionEndHeight := sessionTree.GetSessionHeader().SessionEndBlockHeight
-					currentEndBlockHeight := sessionkeeper.GetSessionEndBlockHeight(block.Height())
-					// TODO_IN_THIS_PR: Add a comment to explain how this check is splitting
-					// the sessions into a list of late session to be claimed ASAP and another
-					// one for the session that also need to be claimed but are on time.
-					if IsWithinGracePeriod(currentEndBlockHeight, sessionEndHeight) {
+					// Separate the sessions that are on-time from the ones that are late.
+					// If the session is past its grace period, it is considered late,
+					// otherwise it is on time and will be emitted last.
+					if IsPastGracePeriod(endBlockHeight, block.Height()) {
 						lateSessions = append(lateSessions, sessionTree)
 					} else {
 						onTimeSessions = append(onTimeSessions, sessionTree)
 					}
 				}
-			}
-		}
 
-		// If there are any late sessions to be claimd, emmit them first.
-		// The wait for claim submission window pipeline step will return without
-		// waiting for block to be committed.
-		if len(lateSessions) > 0 {
-			sessionsToClaimsPublishCh <- lateSessions
+				// If there are any late sessions to be claimed, emit them first.
```

Comment:
I don't understand this. I assumed late sessions are too late and shouldn't be claimed, yet here we are trying to claim them.

This makes me question the rest of the business logic in this function.

Reply:
`IsWithinGracePeriod` is checking for sessions to claim with `<=` operator. Which means that it would catch sessions that were supposed to be claimed in previous block heights too.
These late sessions might have their create claim window closed and are no longer eligible to be claimed, but that's not always the case.
Once claim window closing is implemented, they will be filtered out downstream at the appropriate step.

Will put this explanation as a comment but not at this specific line.